src/map/achievement.c

```diff
@@ -0,0 +1,1693 @@
+#include "map/itemdb.h"
+
+#include "common/memmgr.h"
+#include "common/nullpo.h"
+#include "common/conf.h"
+#include "common/showmsg.h"
+#include "common/strlib.h"
+#include "common/db.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+struct achievement_interface achievement_s;
+struct achievement_interface *achievement;
+
+/**
+ * Retrieve an achievement via it's ID.
+ * @param aid as the achievement ID.
+ * @return NULL or pointer to the achievement data.
+ */
+struct achievement_data *achievement_get(int aid)
+{
+	return (struct achievement_data *) idb_get(achievement->db, aid);
+}
+
+/**
+ * Searches the provided achievement data for an achievement,
+ * optionally creates a new one if no key exists.
+ * @param sd     a pointer to map_session_data.
+ * @param aid    ID of the achievement provided as key.
+ * @param create new key creation flag.
+ * @return pointer to the session's achievement data.
+ */
+struct achievement *achievement_find_or_create(struct map_session_data *sd, int aid, bool create)
+{
+	const struct achievement_data *ad = NULL;
+	struct achievement *a = NULL;
+	int i = 0;
+
+	nullpo_retr(NULL, sd);
+
+	if ((ad = achievement->get(aid)) == NULL)
+		return NULL;
+	
+	/* Lookup for achievement entry */
+	ARR_FIND(0, VECTOR_LENGTH(sd->achievement), i, (a = &VECTOR_INDEX(sd->achievement, i)) && a->id == ad->id);
+
+	if (i == VECTOR_LENGTH(sd->achievement))
+		a = NULL;
+
+	/* Create a new one if required */
+	if (create && a == NULL) {
+		struct achievement ta = { 0 };
+		VECTOR_ENSURE(sd->achievement, 1, 1);
+		ta.id = aid;
+		VECTOR_PUSH(sd->achievement, ta);
+		a = &VECTOR_INDEX(sd->achievement, VECTOR_LENGTH(sd->achievement)-1);
+	}
+
+	return a;
+}
+
+/**
+ * Calculates the achievement's totals via reference.
+ * @param sd         pointer to map_session_data
+ * @param points     pointer to total points var
+ * @param total      pointer to total var
+ * @param completed  pointer to completed var
+ * @param rank       pointer to achievement rank var
+ */
+void achievement_calculate_totals(const struct map_session_data *sd, int *total_points, int *completed, int *rank, int *curr_rank_points)
+{
+	const struct achievement *a = NULL;
+	const struct achievement_data *ad = NULL;
+	int i = 0;
+
+	nullpo_retv(sd);
+	
+	for (i = 0; i < VECTOR_LENGTH(sd->achievement); i++) {
+		a = &VECTOR_INDEX(sd->achievement, i);
+
+		if ((ad = achievement->get(a->id)) == NULL)
+			continue;
+
+		if (a->completed_at != 0 && (total_points != NULL || completed != NULL)) {
+			if (total_points != NULL)
+				*total_points += ad->points;
+			if (completed != NULL)
+				(*completed)++;
+		}
+	}
+
+	if (total_points != NULL && *total_points > 0
+		&& (rank != NULL || curr_rank_points != NULL)) {
+		if (curr_rank_points != NULL )
+			*curr_rank_points = *total_points;
+		for (i = 0; i < MAX_ACHIEVEMENT_RANKS && *curr_rank_points >= VECTOR_INDEX(achievement->rank_exp, i) && i < VECTOR_LENGTH(achievement->rank_exp); i++) {
```

Comment:
Re-coded the function to use temporary buffers and assign the values on non-null pointers  at the end of the function.